File: week12/24/monday/monday.cpp
// Sort Characters By Frequency

#include <string>
#include <unordered_map>
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    string frequencySort(string s) {
        unordered_map<char, int> freq; // 문자와 그 문자의 개수를 저장하는 해시 맵
        for (char c : s) { // s를 순회하면서 각 문자의 개수를 세고, 그 개수와 문자를 튜플로 만들어서 정렬
            freq[c]++; // 문자의 개수를 세고
        }
        vector<pair<char, int>> freq_vec(freq.begin(), freq.end()); // 해시 맵을 벡터로 변환
        sort(freq_vec.begin(), freq_vec.end(), [](const pair<char, int>& a, const pair<char, int>& b) { // a와 b를 비교
            return a.second > b.second; // 문자의 개수를 기준으로 정렬 
        });
        string result; // 결과를 저장할 문자열
        for (const auto& p : freq_vec) { // 정렬된 벡터를 순회하면서 각 문자를 그 문자의 개수만큼 반복하여 문자열로 변환
            result.append(p.second, p.first); // 문자를 그 문자의 개수만큼 반복하여 문자열로 변환
        }
        return result;
    }
};

File: week12/24/monday/monday.py
# Sort Characters By Frequency
from collections import Counter

class Solution:
    def frequencySort(self, s: str) -> str:
        # return ''.join(sorted(s, key=lambda x: (-s.count(x), x)))
        # s를 순회하면서 각 문자의 개수를 세고, 그 개수와 문자를 튜플로 만들어서 정렬
        # 람다 함수는 튜플의 첫번째 요소를 기준으로 정렬하고, 같으면 두번째 요소를 기준으로 정렬
        # ''.join()은 정렬된 문자 리스트를 문자열로 변환
        # LeetCode에서는 Time Limit Exceeded가 발생

        return ''.join([k*v for k, v in Counter(s).most_common()])
        # Counter(s)를 만들어서 각 문자의 개수를 세고, most_common()을 통해 개수가 많은 순서대로 정렬
        # [k*v for k, v in Counter(s).most_common()]은 각 문자와 그 문자의 개수를 곱한 리스트를 만듦
        # ''.join()은 리스트를 문자열로 변환
        # 따라서 각 문자를 그 문자의 개수만큼 반복하여 문자열로 변환

